File: kafkakewl-common/src/main/scala/com/mwam/kafkakewl/common/persistence/KafkaPersistentStore.scala
/*
 * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.common.persistence

import com.mwam.kafkakewl.common.kafka.KafkaClientConfigExtensions.*
import com.mwam.kafkakewl.common.kafka.AdminClientExtensions.*
import com.mwam.kafkakewl.common.kafka.KafkaConsumerUtils
import com.mwam.kafkakewl.domain.*
import com.mwam.kafkakewl.domain.DeploymentsJson.given
import com.mwam.kafkakewl.domain.config.KafkaClientConfig
import zio.kafka.consumer.Consumer
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.config.TopicConfig
import zio.*
import zio.json.*
import zio.kafka.producer.*
import zio.kafka.consumer.*
import zio.kafka.serde.{Deserializer, Serde}
import zio.kafka.admin.*
import zio.kafka.consumer.Consumer.AutoOffsetStrategy
import zio.stream.*

import scala.collection.mutable.ListBuffer

final case class BatchMessageEnvelope[Payload](
    batchSize: Int,
    indexInBatch: Int,
    payload: Payload
)

object BatchMessageEnvelopeJson {
  import zio.json.{DeriveJsonDecoder, DeriveJsonEncoder, JsonDecoder, JsonEncoder}

  given [Payload](using JsonEncoder[Payload]): JsonEncoder[BatchMessageEnvelope[Payload]] = DeriveJsonEncoder.gen[BatchMessageEnvelope[Payload]]
  given [Payload](using JsonDecoder[Payload]): JsonDecoder[BatchMessageEnvelope[Payload]] = DeriveJsonDecoder.gen[BatchMessageEnvelope[Payload]]
}

class KafkaPersistentStore(
    kafkaClientConfig: KafkaClientConfig,
    kafkaPersistentStoreConfig: KafkaPersistentStoreConfig,
    kafkaProducer: TransactionalProducer,
    initializedRef: Ref.Synchronized[Boolean]
) extends PersistentStore {
  import BatchMessageEnvelopeJson.given

  private object KafkaSerde {
    val key: Serde[Any, TopologyId] =
      Serde.string.inmapM(s => ZIO.succeed(TopologyId(s)))(topologyId => ZIO.succeed(topologyId.value))
    val value: Serde[Any, BatchMessageEnvelope[TopologyDeployment]] =
      Serde.string.inmapM[Any, BatchMessageEnvelope[TopologyDeployment]](s =>
        ZIO.fromEither(s.fromJson[BatchMessageEnvelope[TopologyDeployment]]).mapError(e => new RuntimeException(e))
      )(td => ZIO.succeed(td.toJson))
  }

  override def loadLatest(): Task[TopologyDeployments] =
    initializeIfNeeded *>
      ZIO.scoped {
        for {
          consumer <- Consumer.make(
            ConsumerSettings(kafkaClientConfig.brokersList, kafkaClientConfig.additionalConfig).withOffsetRetrieval(
              Consumer.OffsetRetrieval.Auto(AutoOffsetStrategy.Earliest)
            )
          )
          _ <- ZIO.logInfo(s"loading the topologies from the $topicName kafka topic")
          deserializedTopologiesWithDuration <- consumeUntilEnd(consumer, topicName).timed
          (duration, deserializedTopologies) = deserializedTopologiesWithDuration

          failedTopologies = deserializedTopologies.collect { case (id, Left(jsonError)) =>
            ZIO.logError(s"failed to deserialize topology $id: $jsonError")
          }.toList

          topologies = deserializedTopologies
            .collect { case (_, Right(topology)) => topology }
            .map(td => (td.topologyId, td))
            .toMap

          _ <- ZIO.foreachDiscard(failedTopologies)(identity)
          _ <- ZIO.logInfo(
            s"finished loading ${topologies.size} topologies, discarded ${failedTopologies.size} from the $topicName kafka topic in ${duration.toMillis / 1000.0} seconds"
          )
        } yield topologies
      }

  override def save(topologyDeployments: TopologyDeployments): Task[Unit] =
    initializeIfNeeded *> ZIO.scoped {
      for {
        txn <- kafkaProducer.createTransaction
        batchSize = topologyDeployments.size
        _ <- ZIO.foreachDiscard(topologyDeployments.values.zipWithIndex) { (td, indexInBatch) =>
          for {
            recordMetadataWithDuration <- txn
              .produce(
                topicName,
                td.topologyId,
                BatchMessageEnvelope(batchSize, indexInBatch, td),
                KafkaSerde.key,
                KafkaSerde.value,
                offset = None
              )
              .timed
            (duration, recordMetadata) = recordMetadataWithDuration
            _ <- ZIO.logInfo(
              s"saved topology ${td.topologyId} into $topicName kafka topic: P#${recordMetadata.partition} @${recordMetadata.offset} in ${duration.toMillis / 1000.0} seconds"
            )
          } yield ()
        }
      } yield ()
    }

  override def stream(compactHistory: Boolean): Stream[Throwable, TopologyDeployments] =
    ZStream.unwrapScoped {
      initializeIfNeeded *> ZIO.attempt {
        ???
      }
    }

  private def initializeIfNeeded: Task[Unit] = for {
    _ <- initializedRef.updateZIO { initialized =>
      // initialize if it hasn't been and either way set initializedRef to true
      initialize.unless(initialized) *> ZIO.succeed(true)
    }
  } yield ()

  private def initialize: Task[Unit] = {
    ZIO.scoped {
      for {
        adminClient <- AdminClient.make(kafkaClientConfig.toAdminClientSettings)
        topics <- adminClient.listTopics()
        _ <-
          if (topics.contains(topicName)) {
            if (topicConfig.reCreate) {
              recreateTopicZIO(adminClient)
            } else {
              updateTopicZIO(adminClient)
            }
          } else {
            createTopicZIO(adminClient)
          }
      } yield ()
    }
  }

  private def createTopicZIO(adminClient: AdminClient) = {
    val numPartitions = 1
    val replicationFactor = topicConfig.replicationFactor.getOrElse(-1.toShort)
    val config = kafkaTopicConfig
    val newTopic = AdminClient.NewTopic(topicName, numPartitions, replicationFactor, config)
    for {
      _ <- ZIO.logInfo(s"Creating topic $topicName...")
      _ <- adminClient.createTopic(newTopic)
      _ <- ZIO.logInfo(
        s"Topic $topicName created successfully: partitions = $numPartitions, replication-factor = $replicationFactor, config = $config"
      )
    } yield ()
  }

  private def deleteTopicZIO(adminClient: AdminClient) =
    ZIO.logInfo(s"Deleting topic $topicName...") *> adminClient.deleteTopic(topicName)

  private def recreateTopicZIO(adminClient: AdminClient) = {
    val delay = 5.seconds
    // Waiting 5 seconds so that it can be properly deleted (TODO refactor this to a common place)
    deleteTopicZIO(adminClient)
      *> ZIO.logInfo(s"Waiting for ${delay.toSeconds} seconds before creating topic $topicName...")
      *> ZIO.sleep(delay)
      *> createTopicZIO(adminClient)
  }

  private def updateTopicZIO(adminClient: AdminClient) = {
    for {
      _ <- ZIO.logInfo(s"Topic $topicName already exists")
      topicsDescriptions <- adminClient.describeTopics(Seq(topicName))
      topicDescriptions <- ZIO.getOrFailWith(RuntimeException(s"Could not describe topic $topicName"))(topicsDescriptions.get(topicName))

      numPartitions <- ZIO.getOrFailWith(RuntimeException(s"Topic $topicName has no partition infos."))(
        topicDescriptions.partitions.map(_.partition).maxOption.map(_ + 1)
      )
      _ <- ZIO
        .fail(
          RuntimeException(
            s"Topic $topicName has $numPartitions partitions but expected 1. Use a different topic for persistence or delete the $topicName topic and let kafkakewl re-create it with the right number of partitions"
          )
        )
        .unless(numPartitions == 1)

      numReplicas = topicDescriptions.partitions.head.replicas.length
      config <- adminClient.getDynamicTopicConfig(topicName)
      // TODO check replicas
      // TODO check config, alter config if possible (unless the replica constraints have changed)
      _ <- ZIO.logInfo(s"Topic $topicName partitions = $numPartitions, replicas = $numReplicas, config = $config")
    } yield ()
  }

  private def consumeUntilEnd(consumer: Consumer, topic: String): Task[Map[TopologyId, Either[String, TopologyDeployment]]] = {

    for {
      topicPartitions <- consumer.partitionsFor(topic)
      partitionMapAndRecords <- KafkaConsumerUtils.consumeUntilEnd(
        consumer,
        topicPartitions.map(tpi => new TopicPartition(tpi.topic(), tpi.partition())),
        Deserializer.string,
        Deserializer.string
      )
      (partitionMap, records) = partitionMapAndRecords

      topologyMap = records
        .foldLeft(collection.mutable.Map.empty[String, String]) {
          case (map, record) => {
            map.update(record.key(), record.value())
            map
          }
        }
        .map { case (key, value) => (TopologyId(key), value.fromJson[BatchMessageEnvelope[TopologyDeployment]].map(_.payload)) }
        .toMap

    } yield topologyMap
  }
  private def topicConfig = kafkaPersistentStoreConfig.topic
  private def topicName = topicConfig.name
  private val defaultKafkaTopicConfig = Map(TopicConfig.RETENTION_MS_CONFIG -> "-1")
  private def kafkaTopicConfig = defaultKafkaTopicConfig ++ topicConfig.config
}

object KafkaPersistentStore {
  val live: ZLayer[KafkaClientConfig & KafkaPersistentStoreConfig, Throwable, KafkaPersistentStore] =
    ZLayer.scoped {
      for {
        kafkaClientConfig <- ZIO.service[KafkaClientConfig]
        kafkaPersistentStoreConfig <- ZIO.service[KafkaPersistentStoreConfig]
        producer <- TransactionalProducer.make(
          TransactionalProducerSettings(
            kafkaClientConfig.brokersList,
            30.seconds,
            properties = kafkaClientConfig.additionalConfig,
            kafkaPersistentStoreConfig.transactionalId,
            4096
          )
        )
        initialized <- Ref.Synchronized.make(false)
      } yield KafkaPersistentStore(kafkaClientConfig, kafkaPersistentStoreConfig, producer, initialized)
    }
}

File: kafkakewl-common/src/main/kotlin/com/mwam/kafkakewl/common/persistence/KafkaPersistentStore.kt
/*
 * SPDX-FileCopyrightText: 2024 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.common.persistence

import com.mwam.kafkakewl.common.config.KafkaClientConfig
import com.mwam.kafkakewl.common.config.KafkaPersistentStoreConfig
import com.mwam.kafkakewl.common.config.toKafkaClientConfig
import com.mwam.kafkakewl.domain.TopologyDeployment
import com.mwam.kafkakewl.domain.TopologyDeployments
import com.mwam.kafkakewl.domain.TopologyId
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.apache.kafka.clients.admin.AdminClient
import org.apache.kafka.clients.admin.ConfigEntry
import org.apache.kafka.clients.admin.DescribeConfigsOptions
import org.apache.kafka.clients.admin.KafkaAdminClient
import org.apache.kafka.clients.admin.NewTopic
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.config.ConfigResource
import org.apache.kafka.common.config.TopicConfig
import org.apache.kafka.common.serialization.Deserializer
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.kafka.common.serialization.StringSerializer
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

@Serializable
data class BatchMessageEnvelope<Payload>(val batchSize: Int, val indexInBatch: Int, val payload: Payload)

private val logger = KotlinLogging.logger {}

class KafkaPersistentStore(
    private val kafkaClientConfig: KafkaClientConfig,
    private val kafkaPersistentStoreConfig: KafkaPersistentStoreConfig
) : PersistentStore, AutoCloseable {
    companion object {
        private const val TOPIC_NUM_PARTITIONS = 1
    }

    private val producer: KafkaProducer<String, String>

    init {
        val adminConfig = kafkaClientConfig.toKafkaClientConfig()
        KafkaAdminClient.create(adminConfig).use { adminClient ->
            if (adminClient.listTopics().names().get().contains(topic)) {
                // The topic already exists
                if (kafkaPersistentStoreConfig.topic.reCreate) {
                    reCreateTopic(adminClient)
                } else {
                    checkExistingTopic(adminClient)
                }
            } else {
                // The topic does not exist yet
                createTopic(adminClient)
            }
        }

        val producerConfig = adminConfig + mapOf(ProducerConfig.TRANSACTIONAL_ID_CONFIG to kafkaPersistentStoreConfig.transactionalId)
        producer = KafkaProducer<String, String>(producerConfig, StringSerializer(), StringSerializer())
        producer.initTransactions()
    }

    override fun loadLatest(): Result<TopologyDeployments> {
        return runCatching {
            val (topologyDeployments, _) = loadMessagesUntilEnd<TopologyDeployments, String, String>(
                kafkaClientConfig,
                topic,
                StringDeserializer(),
                StringDeserializer(),
                { emptyMap() },
                // TODO json deserialization errors should include offset or the json itself
                { msgs, cr -> msgs + (TopologyId(cr.key()) to Json.decodeFromString<BatchMessageEnvelope<TopologyDeployment>>(cr.value()).payload) },
            )

            topologyDeployments
        }
    }

    override fun save(topologyDeployments: TopologyDeployments): Result<Unit> {
        return runCatching {
            try {
                producer.beginTransaction()
                val batchSize = topologyDeployments.size
                topologyDeployments.entries
                    .forEachIndexed { indexInBatch, entry ->
                        val key = entry.key.value
                        val value = Json.encodeToString(BatchMessageEnvelope(batchSize, indexInBatch, entry.value))
                        val record = ProducerRecord(topic, key, value)
                        // No callback, in case of failure, commitTransaction will throw, and we'll get the exception anyway
                        producer.send(record)
                    }
                producer.commitTransaction()
            } catch (e: Exception) {
                producer.abortTransaction()
                throw e
            }
        }
    }

    override fun stream(compactHistory: Boolean): Flow<TopologyDeployments> {
        TODO("not implemented")
    }

    override fun close() {
        producer.close()
    }

    private val topic: String
        get() = kafkaPersistentStoreConfig.topic.name

    private val replicationFactor: Short
        get() = kafkaPersistentStoreConfig.topic.replicationFactor ?: -1

    private val topicConfig: Map<String, String>
        get() = kafkaPersistentStoreConfig.topic.config + (TopicConfig.RETENTION_MS_CONFIG to "-1")

    private fun createTopic(adminClient: AdminClient) {
        val newTopic = NewTopic(topic, TOPIC_NUM_PARTITIONS, replicationFactor).configs(topicConfig)
        logger.info { "Creating topic $topic with partitions = $TOPIC_NUM_PARTITIONS, replicationFactor = $replicationFactor, config = $topicConfig ..." }
        adminClient.createTopics(listOf(newTopic)).all().get()
        logger.info { "Finished creating topic $topic" }
    }

    private fun checkExistingTopic(adminClient: AdminClient) {
        val topicDescription = adminClient.describeTopics(listOf(topic)).allTopicNames().get()[topic]!!
        val actualNumPartitions = topicDescription.partitions().map { it.partition() }.max() + 1
        if (actualNumPartitions != TOPIC_NUM_PARTITIONS) {
            throw RuntimeException("Topic $topic has $actualNumPartitions partitions but expected $TOPIC_NUM_PARTITIONS. Use a different topic for persistence or delete the $topic topic and let kafkakewl re-create it with the right number of partitions")
        }
        val actualNumReplicas = topicDescription.partitions().first().replicas().size.toShort()
        if (replicationFactor != (-1).toShort() && actualNumReplicas != replicationFactor) {
            throw RuntimeException("Topic $topic has $actualNumReplicas replicas but expected $replicationFactor. Use a different topic for persistence or delete the $topic topic and let kafkakewl re-create it with the right number of partitions")
        }
        val topicConfigResource = ConfigResource(ConfigResource.Type.TOPIC, topic)
        val actualConfig = adminClient.describeConfigs(listOf(topicConfigResource), DescribeConfigsOptions()).all().get()[topicConfigResource]!!
            .entries()
            .filter { it.source() == ConfigEntry.ConfigSource.DYNAMIC_TOPIC_CONFIG }
            .associate { it.name()!! to it.value()!! }
        if (actualConfig != topicConfig) {
            throw RuntimeException("Topic $topic has $actualConfig config but expected $topicConfig. Use a different topic for persistence or delete the $topic topic and let kafkakewl re-create it with the right number of partitions")
        }
    }

    private fun reCreateTopic(adminClient: AdminClient) {
        deleteTopic(adminClient)
        createTopic(adminClient)
    }

    private fun deleteTopic(adminClient: AdminClient) {
        logger.info { "Deleting topic $topic ..." }
        adminClient.deleteTopics(listOf(topic)).all().get()
        val delay = 5.seconds.toJavaDuration()
        logger.info { "Finished deleting $topic, waiting $delay so that it's properly deleted" }
        Thread.sleep(delay)
    }

    private fun <Messages, K, V> loadMessagesUntilEnd(
        kafkaClientConfig: KafkaClientConfig,
        topic: String,
        keyDeserializer: Deserializer<K>,
        valueDeserializer: Deserializer<V>,
        initialMessages: () -> Messages,
        combine: (Messages, ConsumerRecord<K, V>) -> Messages,
        pollTimeout: Duration = 100.milliseconds,
    ): Pair<Messages, Map<TopicPartition, Long>> {
        val consumerConfig = kafkaClientConfig.toKafkaClientConfig() + mapOf(
            ConsumerConfig.AUTO_OFFSET_RESET_CONFIG to "earliest",
            ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG to false
        )
        val pollTimeoutJava = pollTimeout.toJavaDuration()
        KafkaConsumer<K, V>(consumerConfig, keyDeserializer, valueDeserializer).use { consumer ->
            val topicPartitions = consumer.partitionsFor(topic).map { TopicPartition(it.topic(), it.partition()) }
            val endOffsets = consumer.endOffsets(topicPartitions).map { it.key!! to it.value!! }.toMap()

            var positions = emptyMap<TopicPartition, Long>()
            var messages = initialMessages()
            consumer.assign(topicPartitions)
            do {
                val consumerRecords = consumer.poll(pollTimeoutJava)
                messages = consumerRecords.fold(messages, combine)
                positions = endOffsets.mapValues { consumer.position(it.key) }
                val hasNotReachedTheEnd = endOffsets.any { it.value > positions.getValue(it.key) }
            } while (hasNotReachedTheEnd)
            return messages to positions
        }
    }
}
